### PR TITLE
[opentitanlib] Add strappings for hyper310 SPI TPM

### PR DESCRIPTION
**Pull request by a-will**
The "QSPI" SPI interface is mapped to the STM32's OCTOSPI peripheral in the hyperdebug firmware, but it is incapable of performing the TPM protocol. Use SPI1 instead, and add strappings to set pins to high impedance for the unused interface, since they will both be driving the same traces.

### CODE REVIEW

**Comment by a-will** [comment]
What's not addressed here is what we want to do with CW310 + hyperdebug in CI. Do we respin the swizzle board? Or merely use jumper wire to connect the required pins?

CW340 would also need jumper wires to connect pins (but it would use the top of the nucleo board instead of the bottom).

**Review by engdoreis — APPROVED** [review]

**Comment by engdoreis** [comment]
Regarding CI, I think we should respin the swizzle board in order to avoid flakiness caused by loose cables, as we had recently with hand made ribbon cables.

**Comment by a-will** [comment]
> Regarding CI, I think we should respin the swizzle board in order to avoid flakiness caused by loose cables, as we had recently with hand made ribbon cables.

Unfortunately, I don't think there is any mechanical support for something more secure on the CW310. These don't seem like great connectors for board-to-board mating, but I'm also not a board designer, so I should defer to an expert here. Maybe @colinoflynn would know whether the approach could be improved.

**Comment by a-will** [comment]
I suppose for this patch, we don't have to solve the hardware problem before it can be merged. We'll be unable to run tests in CI until we have the hardware, though.

Offline, @mazurek-michal and I have at least confirmed SPI TPM communication works when this strapping is applied.

**Comment by colinoflynn** [comment]
> Regarding CI, I think we should respin the swizzle board in order to avoid flakiness caused by loose cables

The connectors can be used with "standard" 0.05" female headers, although they are designed for mating with IDC cables. The original idea was something like 20021321-00040T4LF+ 20021321-00034T4LF could be used (40-pin & 34-pin). There are some examples on the (Mating Headers)[https://rtfm.newae.com/Targets/CW310%20Bergen%20Board/#mating-headers] section but they haven't been checked physically.

**Comment by a-will** [comment]
> > Regarding CI, I think we should respin the swizzle board in order to avoid flakiness caused by loose cables
> 
> The connectors can be used with "standard" 0.05" female headers, although they are designed for mating with IDC cables. The original idea was something like 20021321-00040T4LF+ 20021321-00034T4LF could be used (40-pin & 34-pin). There are some examples on the (Mating Headers)[https://rtfm.newae.com/Targets/CW310%20Bergen%20Board/#mating-headers] section but they haven't been checked physically.

Yeah, keeping IDC cables seemed like the simple path forward, especially since the flexibility of cables is probably desirable with what the CW310 provides.

After having sheared off part of a connector housing, I'm probably excessively concerned about such things, but I don't see any supporting structure for a board to mount on top of those connectors, for example. It's not like we've got something like FMC's standoffs, and mating two PCBs directly would mean needing to put some effort into the mechanical dimensions for the assembly (matching connector spacing and board heights and such).

**Comment by luismarques** [comment]
> Upon further consideration, I believe that soldering wires on the swizzle board would not be a bad idea in terms of reliability. However, it would be bad for reproducibility, as one cannot simply purchase the boards and cables to replicate the CI environment.

We can solder the wires for now and come up with a longer term solution with more time.

**Comment by charles-at-lowrisc** [comment]
Is it possible to make the failure of the test report to the user that the hardware patch may not have been applied or might have failed, so that others who try to reproduce CI at least know that it's not their fault if it comes up.

And update documentation so that others outside the company know there's an expected failure if that change isn't made on the board, and how to address it? (feels ugly, but I suspect that this sort of thing is probably common as a hardware errata?)

**Comment by colinoflynn** [comment]
Or a 0.05" breakout, which would make soldering easier, and for someone else building up this board could be made available (e.g., all the files posted)? Not sure on timeline but that would obviously be very fast!

I think the original thought-process on my side was that we might build a custom STM32L552 debug probe plugging in there, so you wouldn't need the larger size of the Nucleo board. Since the days of STM32 drought we had a bunch of STM32L552ZET6 sitting in inventory in case that was needed...

**Comment by jesultra** [comment]
Soldering three wires on top of a Swizzle board sounds like a fine solution for a few CI setups.  Then ideally, the change would be incorporated into a future revision of the Swizzle board.

It is very possibe to craft a test to verify the presence of the required connections.  We are talking about three wires that each join together pairs of HyperDebug pins.  HyperDebug could be instructed to put one pin from each pair into Input mode, and the three other pins into PushPull mode, then one of the three pins high at a time, each time verifying that the correct input pin follows along.  That would instantly detect any missing or crossed wires.

As long as the above is done either while the OpenTitan RESET signal is asserted, or at least while none of the chip select signals are asserted, then driving the clock and data lines of a SPI bus will not cause a drive fight between HyperDebug and CW310.

**Pull request merged**
